dataflow/src/domain/mod.rs

```diff
@@ -2453,44 +2494,63 @@ impl Domain {
                             let ref n = *nd.borrow();
                             let local_index: LocalNodeIndex = *n.local_addr();
 
-                            self.state
-                                .get(&local_index)
-                                .filter(|state| state.is_partial())
-                                .map(|state| (local_index, state.deep_size_of()))
+                            if n.is_reader() {
+                                let mut size = 0;
+                                n.with_reader(|r| size = r.state_size().unwrap_or(0))
+                                    .unwrap();
+                                Some((local_index, size))
+                            } else {
+                                self.state
+                                    .get(&local_index)
+                                    .filter(|state| state.is_partial())
+                                    .map(|state| (local_index, state.deep_size_of()))
+                            }
                         })
                         .filter(|&(_, s)| s > 0)
                         .max_by_key(|&(_, s)| s)
                         .map(|(n, s)| {
-                            warn!(self.log, "chose to evict from node {:?} with size {}", n, s);
-                            n
+                            trace!(self.log, "chose to evict from node {:?} with size {}", n, s);
+                            (n, cmp::min(num_bytes, s as usize))
                         })
                 });
 
-                if let Some(node) = node {
+                if let Some((node, num_bytes)) = node {
                     let mut freed = 0u64;
                     while freed < num_bytes as u64 {
                         use core::data::SizeOf;
 
-                        let (key_columns, keys, bytes) = {
-                            let k = self.state[&node].evict_random_keys(100);
-                            (k.0.to_vec(), k.1, k.2)
-                        };
-                        freed += bytes;
+                        if self.nodes[&node].borrow().is_reader() {
+                            // we can only evict one key a time here because the freed memory
+                            // calculation is based on the key that *will* be evicted. We may count
+                            // the same individual key twice if we batch evictions here.
```

Comment:
I'm not sure I understand this logic? What is stopping us from clearing multiple buckets in the `Reader` at one time?

Reply:
See the comment. We may choose an index that maps to the same bucket, and double-count it.

Reply:
If `WriteHandle` had a `evict_random_keys`, it could just make sure it doesn't do that in the loop in https://github.com/mit-pdos/distributary/pull/75/files#diff-117bbb5259149102b8211b4567837f17R250 no?